### PR TITLE
Factor out particle frame list <-> particle array into PMacc

### PR DESCRIPTION
**Pull request by Flamefire**
This moves the particle linked frame list to contiguous array conversion into PMacc as this is also required for other libraries doing check-pointing etc. There were some minor adjustments so the methods don't rely on any data from the user library except the arguments.

I would even go further for the split* function and create a wrapper in PMacc that also does the `counterBuffer` creation and chunk-wise loading as done in HDF5/ADIOS LoadSpecies (the code there is a complete duplicate except the log entries) @ax3l ?
ccing @psychocoderHPC

### CODE REVIEW

File: src/libPMacc/include/particles/operations/ConcatListOfFrames.hpp [hunk 1]
```diff
@@ -0,0 +1,143 @@
+/**
+ * Copyright 2013-2016 Rene Widera, Felix Schmitt, Alexander Grund
+ *
+ * This file is part of PIConGPU.
+ *
+ * PIConGPU is free software: you can redistribute it and/or modify
```
Comment (psychocoderHPC) [inline src/libPMacc/include/particles/operations/ConcatListOfFrames.hpp]: Please change the license to the PMacc GPL + LPGL

Comment (Flamefire) [inline src/libPMacc/include/particles/operations/ConcatListOfFrames.hpp]: Done

**Comment by ax3l** [comment]
thanks for the refactoring!

> I would even go further for the split* function and create a wrapper in PMacc that also does the counterBuffer creation and chunk-wise loading as done in HDF5/ADIOS LoadSpecies (the code there is a complete duplicate except the log entries) @ax3l ?

yes, we can do this as a follow-up. take care, I am currently modifying the HDF5 writer a bit.

File: src/libPMacc/include/particles/operations/splitIntoListOfFrames.kernel [hunk 2]
```diff
@@ -21,30 +21,34 @@
```
Comment (psychocoderHPC) [inline src/libPMacc/include/particles/operations/splitIntoListOfFrames.kernel]: Please update the license file. GPL->LGPL

File: src/libPMacc/include/particles/operations/ConcatListOfFrames.hpp [hunk 3]
```diff
@@ -0,0 +1,145 @@
+/**
+ * Copyright 2013-2016 Rene Widera, Felix Schmitt, Alexander Grund
```
Comment (psychocoderHPC) [inline src/libPMacc/include/particles/operations/ConcatListOfFrames.hpp]: @f-schmitt-zih Are you fine with this license change from GPL to LGPL?

Comment (f-schmitt-zih) [inline src/libPMacc/include/particles/operations/ConcatListOfFrames.hpp]: Sounds good to me

**Comment by Flamefire** [comment]
Updated license and moved also some more stuff to PMacc so you don't have to create the counterBuffer

**Comment by ax3l** [comment]
-> merge after 0.2.0 release branch has been opened

**Comment by Flamefire** [comment]
@ax3l Any chance this gets into dev soon? paraTAXIS depends on this, so would be great :)

**Comment by psychocoderHPC** [comment]
I am not @ax3l, but yes we can get it in soon. We branched the release and now only port bugs back.

File: src/libPMacc/include/particles/operations/splitIntoListOfFrames.kernel [hunk 4]
```diff
@@ -0,0 +1,214 @@
...
+             );
+        destSpecies.fillAllGaps();
+        leftOverParticles -= currentChunkSize;
+    }
+
+    counterBuffer.deviceToHost();
+    log(logLvl, "wait for last processed chunk: %1%") % T_SrcFrame::getName();
+    __getTransactionEvent().waitForFinished();
+
+    log<PMaccVerbose::INFO>("used frames to load particles: %1%") % counterBuffer.getHostBuffer().getDataBox()[2];
```
Comment (psychocoderHPC) [inline src/libPMacc/include/particles/operations/splitIntoListOfFrames.kernel]: Please pass the error/info to `logLvl` instead of `PMaccVerbose::INFO`

**Comment by psychocoderHPC** [comment]
@Flamefire  Do you tested if this pull request runs with adios?

File: src/picongpu/include/plugins/hdf5/restart/LoadSpecies.hpp [hunk 5]
```diff
@@ -202,40 +202,9 @@ struct LoadSpecies
 
             const uint32_t cellsInSuperCell = PMacc::math::CT::volume<SuperCellSize>::type::value;
 
-            const uint32_t iterationsForLoad = ceil(float_64(totalNumParticles) / float_64(restartChunkSize));
```
Comment (psychocoderHPC) [inline src/picongpu/include/plugins/hdf5/restart/LoadSpecies.hpp]: Please remove `counterBuffer` and the comment. Some lines above this code (line 192 - 201)

**Comment by Flamefire** [comment]
Tested restart after 1000 timesteps with ADIOS. Total run time 2000 timesteps, no difference in numerical heating 
figure_1

**Pull request merged**